New beatmap carousel design implementation

mk56-spn opened the pull request:
A few days late but here's some more https://github.com/ppy/osu/issues/18890 stuff...

Some things of note: 

- The  coloured ```starcounter``` here and in  https://github.com/ppy/osu/pull/22116 can be merged into a shared implementation once both PRs are in 

- The outer faded border present in the Figma design on the selected carouselbeatmap has not been added due to  the problems with aliasing discussed here: 
       https://discord.com/channels/188630481301012481/188630652340404224/1063458499725643898.

    I wanted confirmation that its OK to add another nested container to simulate the effect without having this issue since performance considerations might be a thing for the carousel  

- The design uses white for the  higher end of the star rating pill and Ruleset icon in the carousel , despite using yellow everywhere else, should this be done or no? 

![img](https://user-images.githubusercontent.com/74463310/212489059-2b6db797-e7fd-43f0-98fa-b877fd4f40b2.png)

- as with   https://github.com/ppy/osu/pull/22116 there is a gradient present on the CarouselBeatmapHeader. but no explanation on  figma of how to obtain said colour, one has been requested: https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P?node-id=2980:3361#340801912

Preview:
 
![osu_2023-09-08_21-58-01](https://github.com/ppy/osu/assets/35318437/7ef909ec-a3a1-4e26-870a-02e558c3e25a)

As for why i have implemented this directly instead of in a test scene as with the wedge and footer, well the carousel interacts with too many stuff and it would require doubling up on an unreasonable amount of files so i opted for direct integration, even if it means I suffer through a couple of merge conflicts

ps. header coming soon(tm) once i figure out the whole colour updating thing discussed here: https://discord.com/channels/188630481301012481/188630652340404224/1063589348265693204

mk56-spn committed:
Remove old elements, add new colours, reshuffle order of some element in drawable

mk56-spn committed:
Resize and organise elements in DrawableCarouselBeatmap.cs

mk56-spn committed:
Add new icon design to DrawableCarouselBeatmap.cs

mk56-spn committed:
add difficulty spectrum to DrawableCarouselBeatmapSet.cs and adjust rank size on DrawableCarouselBeatmap.cs

mk56-spn committed:
Touch up designs, add outer coloured glow

mk56-spn committed:
Add a small annotation

mk56-spn commented:
went over the test failures, despite every single one managing to fail all of them seem to be the usual suspects unrelated to my changes

bdach commented:
> went over the test failures, despite every single one managing to fail all of them seem to be the usual suspects unrelated to my changes

Clearly not. `TestDifficultyIconSelecting` is exercising the behaviour of being able to select a difficulty clicking on the old difficulty icons (and this also includes switching ruleset). You've removed it by migrating to `DifficultySpectrumDisplay` which does not support this.

Now we may *want* to get rid of that behaviour (it was a point of frequent confusion for users), but that should have had preceding discussion, and if accepted, should be handled by removing the broken tests.

mk56-spn commented:
right my bad didnt make the connection will go over that, and well about preceeding dicussion i thought the design having the difficulty spectrum served as implicit acceptance that this is the way forward, but i guess not

bdach commented:
@ppy/area-client in your view, should the behaviour of clicking on difficulty icons on the carousel panels navigating to said difficulty be retained with this design?

Joehuu commented:
> @ppy/area-client in your view, should the behaviour of clicking on difficulty icons on the carousel panels navigating to said difficulty be retained with this design?

I think to some degree. I still want to change rulesets if a beatmap had multiple. If it was just changing difficulty, I would just click the panels as they are easier to click.

mk56-spn commented:
as an aside https://discord.com/channels/188630481301012481/188630652340404224/1063846141260746843  it seems my little attempt at circumventing colour bleeding around the main colour of the carousel header doesn't work when opacity is involved, given this is the case shall i just let the colour bleed a little or attempt to use buffered containers and incur the performance hit, not sure how big it is so im asking just incase

smoogipoo commented:
> in your view, should the behaviour of clicking on difficulty icons on the carousel panels navigating to said difficulty be retained with this design?

My opinion has unchanged - I don't like having auxiliary methods to do things as it increases cognitive load and UX-wise smaller elements are hard to aim at. Plus, click-object-inside-click-object and scroll-field-inside-scroll-field feels shocking to use.

peppy commented:
It doesn't make any sense with the new design. Removing is fine for now.

mk56-spn committed:
change mapper osuspritetext to use colourprovider colours

mk56-spn committed:
Remove invalid difficulty icon tests

mk56-spn committed:
Merge branch 'master' into beatmap_carousel_new_clean

mk56-spn committed:
Fix issue with panel content

peppy commented:
![osu! 2023-01-17 at 04 18 29](https://user-images.githubusercontent.com/191335/212809106-c46418f2-043f-484a-8619-3d9268000846.png)

1: What's going on here? In the design, there's a shadow layer under each panel, but visually it seems like there's a misalignment only for the selected panel. is this intentional? Is the shadow layer intentionally missing? What is this even @arflyte are you misaligning elements or is it intentional? @arflyte can you please document your designs better so we don't have to go through this every time?

2: Will require attention. The previous implementation used buffering for the background layer, something similar may need to be applied here.

![osu! 2023-01-17 at 04 23 53](https://user-images.githubusercontent.com/191335/212809632-d5c5145c-0b65-446c-9f2d-b9eebe510a84.png)

Glow and gradient border colours don't match new design and feel off.

peppy commented:
Unrelated but `ConstrainedIconContainer` cuts off edge pixels of icons and makes them look very shoddy. Noticed in another recent PR, but now that it's being used in every carousel panel it's a lot more obvious.

peppy reviewed (commented):
Have not checked code, just design and feel

arflyte commented:
![Figma_230117-143706](https://user-images.githubusercontent.com/3632750/212818442-75118836-d785-424a-a442-7a1670f44e7a.png)
![Figma_230117-143731](https://user-images.githubusercontent.com/3632750/212818447-e75deb9a-b97b-486e-b951-1dcb02d51a86.png)

The design is intentional and not misalignment; it's to highlight the selected item better. Also I wanted the selected item to glow, hence the shadow is removed (coz glow and shadow at the same time is kinda weird).

mk56-spn commented:
> Unrelated but `ConstrainedIconContainer` cuts off edge pixels of icons and makes them look very shoddy. Noticed in another recent PR, but now that it's being used in every carousel panel it's a lot more obvious.

is the implying i should remove it or just that it needs to be fixed elsewhere?

mk56-spn commented:
> ![osu! 2023-01-17 at 04 18 29](https://user-images.githubusercontent.com/191335/212809106-c46418f2-043f-484a-8619-3d9268000846.png)
> 
> 1: What's going on here? In the design, there's a shadow layer under each panel, but visually it seems like there's a misalignment only for the selected panel. is this intentional? Is the shadow layer intentionally missing? What is this even @arflyte are you misaligning elements or is it intentional? @arflyte can you please document your designs better so we don't have to go through this every time?
> 
> 2: Will require attention. The previous implementation used buffering for the background layer, something similar may need to be applied here.
> 
> ![osu! 2023-01-17 at 04 23 53](https://user-images.githubusercontent.com/191335/212809632-d5c5145c-0b65-446c-9f2d-b9eebe510a84.png)
> 
> Glow and gradient border colours don't match new design and feel off.

will go over these in detail

mk56-spn committed:
add buffered container to background to allow for uniform opacity

mk56-spn commented:
Going to go over this one tomorrow since I want to adjust a few things with the new info I have from here. Just making it clear that I'm not done addressing the faults that have been brought up.

peppy commented on osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:
```diff
@@ -102,37 +99,26 @@ private void load(BeatmapManager? manager, SongSelect? songSelect)
-                            Width = 40,
-                        },
-                        colourUnderline = new Box
+                        colourBox = new Box { RelativeSizeAxes = Axes.Both, },
```
This starts as a very bright white before difficulty calculation kicks in or whatever. Probably should be something more neutral like a grey?

mk56-spn replied:
right, and perhaps some easing im thinking?

peppy commented on osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:
```diff
@@ -102,37 +99,26 @@ private void load(BeatmapManager? manager, SongSelect? songSelect)
             Header.Children = new Drawable[]
             {
-                new Container
+                new BufferedContainer
```
![osu! 2023-01-23 at 04 54 27](https://user-images.githubusercontent.com/191335/213968248-9db84e07-1f66-4442-b4f8-c8f85b56056d.png)

Probably want to avoid this bleed?

mk56-spn committed:
add pixel hack to avoid colour bleed on carousel

mk56-spn committed:
Merge branch 'master' into beatmap_carousel_new_clean

mk56-spn committed:
Fix leftover unused Drawable from merge

peppy committed:
Merge branch 'master' into beatmap_carousel_new_clean

peppy committed:
Remove unused `DifficultyIcon` implementations

peppy committed:
Add basic test scene to `TestSceneBeatmapCarousel`

mk56-spn commented:
this branch has issues btw, i just dont know how to adress them, namely the alisaing on the transparent corner radius, i will give it a pass in the coming days if you want to wait

peppy commented:
How were you even developing this? There's no usable test scene.

Joehuu committed:
Fix param rename after merge

Joehuu committed:
Cache colour provider screen/test-wide

We may want the plum colour in song select when playing multiplayer/playlists.

Joehuu committed:
Add some way to visualize difficulty colors in test

Similar to what was done in new wedge design PR.

Joehuu committed:
Set 0.00 difficulty color for initial display

Joehuu committed:
Use colour underline again instead of padding hack

Joehuu committed:
Fix beatmap set header border not being "inner alignment"

As said on Figma. Inner alignment can be done by separating the content with containers. Not sure about the performance implications. It also removes the horrid hover layer workaround.

It was outer alignment before, but something about the transparency is making the corner radius aliased. The difficulty drawables need it but is disabled right now.

Joehuu committed:
Make some hardcoded numbers consts

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22197

Joehuu commented:
Have fixed most of the visual problems with slight code quality.

I haven't done the splitting difficulty panel as I still don't really understand the `BeatmapCarousel` code in general after a brief look at it (most importantly the drawable and non-drawable counterparts).

Joehuu committed:
Fix missing initial color of colour underline

Joehuu committed:
Change gradient of set header border to horizontal

Joehuu committed:
Update makeshift border of difficulty panels

Joehuu committed:
Add star rating gradient and slight background gradient

Left gradient opacity is toned down a little bit.

Joehuu committed:
Remove unnecessary fill flow

Joehuu committed:
Add shear to carousel panels

Joehuu committed:
Add right arrow to set panel when selected

Joehuu committed:
Add experimental right arrow animations to hide fade overlap imperfections

Fine with this being tweaked or not shipped at all.

Joehuu committed:
Improve edge effect of difficulty panels

Joehuu committed:
Rename `HasBorder` to `HasCustomBorder`

Joehuu committed:
Improve/update edge effect of set panels

Joehuu committed:
Add comment explaining colour sampling of set panel border and arrow

Joehuu committed:
Move const variables above

Joehuu commented:
Have updated the design:

https://github.com/ppy/osu/assets/35318437/8ddd68cc-4c3b-4677-a2e0-3e5e126d98d5

The code quality has taken a hit as existing components are not flexible and would need a refactor. Let me know if it's worth it (i.e. abstracting `CarouselHeader`).

peppy commented:
![2023-09-10 03 04 00@2x](https://github.com/ppy/osu/assets/191335/4f1ba0b1-d619-437b-89c1-31cb1d25a57a)

something is wrong with this (the gray weird on the edge)

peppy commented on osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:
```diff
@@ -212,10 +264,40 @@ protected override void ApplyState()
+                    {
+                        Type = EdgeEffectType.Shadow,
+                        Colour = starCounter.Colour.MultiplyAlpha(0.5f),
+                        Radius = 50
```
This might be a performance issue, and i'm not sure how good this looks visually either.

peppy committed:
Adjust animations

peppy commented:
I'd probably except the caret arrow to move with the animation rather than just fade in static.

Joehuu committed:
Adjust some metrics to match figma

Joehuu committed:
Remove most aliased edges

The lines added are from trial and error. Apparently this has to be checked with all UI scales.

Joehuu committed:
Rewrite colour provider todo to be more understandable

Joehuu committed:
Add placeholder background to hide left arrow content when beatmap has no background or is loading

Joehuu commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -29,6 +29,10 @@ public partial class PlaySongSelect : SongSelect
     {
         private OsuScreen? playerLoader;
 
+        // TODO: change test colour to a more unique one
```
Have rewritten it.
```
// TODO: pick a different colour as this is using the same colour used in all figma designs
```

Joehuu committed:
Fix crash due to header children load changing

Joehuu commented:
Will try to update this with single diff display throughout the weekend.

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22197

Joehuu committed:
Simplify border edge effect

- No more border gradient (probably same as https://github.com/ppy/osu/pull/22237#issuecomment-2089402615)
- No more radius 50 which is maybe a performance issue

Joehuu committed:
Add single beatmap difficulty display

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22197

Joehuu committed:
Fix diff count tests

Joehuu commented:
Pushing with whatever progress I made in the past.

Haven't addressed the dynamic shearing as commented in https://discord.com/channels/188630481301012481/188630652340404224/1182168002741948426, as it may be involved (i.e. to get the fancy dynamic shear fill flow, you would need to negate the dynamic shear on every child). I would just go with no shear for now, and do a follow-up when adding customisation options.

This probably needs to be rewritten as new `V2` components, but I don't know where to start (i.e. continue this PR or make a fresh PR?). I would prefer a fresh PR, but just asking before I do it. Also, can reconsider unassigning myself to someone with more time/competence based on [this discord message](https://discord.com/channels/188630481301012481/188630652340404224/1263052800540278886), so I can focus more on the wedge.

peppy committed:
Merge branch 'master' into beatmap_carousel_new_clean
